src/EditorFeatures/Core/Implementation/CommentSelection/AbstractCommentSelectionCommandHandler.cs

```diff
@@ -0,0 +1,156 @@
+                if (document == null)
+                {
+                    return true;
+                }
+
+                var service = GetService(document);
+                if (service == null)
+                {
+                    return true;
+                }
+
+                var trackingSpans = new List<CommentTrackingSpan>();
+                var textChanges = new List<TextChange>();
+                CollectEdits(
+                    document, service, textView.Selection.GetSnapshotSpansOnBuffer(subjectBuffer),
+                    textChanges, trackingSpans, operation, CancellationToken.None);
+
+                using (var transaction = new CaretPreservingEditTransaction(title, textView, _undoHistoryRegistry, _editorOperationsFactoryService))
+                {
+                    document.Project.Solution.Workspace.ApplyTextChanges(document.Id, textChanges.Distinct(), CancellationToken.None);
+                    transaction.Complete();
+                }
+
+                using (var transaction = new CaretPreservingEditTransaction(title, textView, _undoHistoryRegistry, _editorOperationsFactoryService))
+                {
+                    Format(service, subjectBuffer.CurrentSnapshot, trackingSpans, CancellationToken.None);
+                    transaction.Complete();
+                }
+
+                if (trackingSpans.Any())
+                {
+                    SetTrackingSpans(textView, subjectBuffer, trackingSpans);
+                }
+            }
+
+            return true;
+        }
+
+        private static ICommentSelectionService GetService(Document document)
+        {
+            // First, try to get the new service for comment selection.
+            var service = document.GetLanguageService<ICommentSelectionService>();
+            if (service != null)
+            {
+                return service;
+            }
+
+            // If we couldn't find one, fallback to the legacy service.
+#pragma warning disable CS0618 // Type or member is obsolete
+            var legacyService = document.GetLanguageService<ICommentUncommentService>();
+#pragma warning restore CS0618 // Type or member is obsolete
+            if (legacyService != null)
+            {
+                return new CommentSelectionServiceProxy(legacyService);
+            }
+
+            return null;
+        }
+
+        public abstract string DisplayName { get; }
+
+        protected abstract string GetTitle(Operation operation);
+
+        protected abstract string GetMessage(Operation operation);
+
+        protected abstract void SetTrackingSpans(ITextView textView, ITextBuffer buffer, List<CommentTrackingSpan> trackingSpans);
+
+        internal abstract void CollectEdits(
+            Document document, ICommentSelectionService service, NormalizedSnapshotSpanCollection selectedSpans,
+            List<TextChange> textChanges, List<CommentTrackingSpan> trackingSpans, Operation operation, CancellationToken cancellationToken);
```

Comment:
why internal?

Reply:
The existing test class for comment selection assumes this is internal.  Left it as-is so I don't need to refactor those tests.

Reply:
wait... we have tests that call directly into this?  ugh...   

Ok, that's terrible.  but you don't have to change.  Just put a comment for now saying `// internal for testing purposes`.  I'll see if i can find time ever to make it so that the tests don't call internal details, but just access the external API of the command-handling system.